legacy/firmware/layout2.c

```diff
@@ -423,18 +423,66 @@ void layoutConfirmOpReturn(const uint8_t *data, uint32_t size) {
                     NULL);
 }
 
-void layoutConfirmTx(const CoinInfo *coin, uint64_t amount_out,
-                     uint64_t amount_fee) {
-  char str_out[32] = {0}, str_fee[32] = {0};
-  bn_format_uint64(amount_out, NULL, coin->coin_shortcut, coin->decimals, 0,
-                   false, str_out, sizeof(str_out));
-  bn_format_uint64(amount_fee, NULL, coin->coin_shortcut, coin->decimals, 0,
-                   false, str_fee, sizeof(str_fee));
+static bool formatAmountDifference(const CoinInfo *coin, uint64_t amount1,
+                                   uint64_t amount2, char *output,
+                                   size_t output_length) {
+  uint64_t abs_diff = 0;
+  const char *sign = NULL;
+  if (amount1 >= amount2) {
+    abs_diff = amount1 - amount2;
+  } else {
+    abs_diff = amount2 - amount1;
+    sign = "-";
+  }
+
+  return bn_format_uint64(abs_diff, sign, coin->coin_shortcut, coin->decimals,
+                          0, false, output, output_length) != 0;
+}
+
+void layoutConfirmTx(const CoinInfo *coin, uint64_t total_in,
+                     uint64_t total_out, uint64_t change_out) {
+  char str_out[32] = {0};
+  formatAmountDifference(coin, total_in, change_out, str_out, sizeof(str_out));
+
+  char str_fee[32] = {0};
+  formatAmountDifference(coin, total_in, total_out, str_fee, sizeof(str_fee));
+
   layoutDialogSwipe(&bmp_icon_question, _("Cancel"), _("Confirm"), NULL,
                     _("Really send"), str_out, _("from your wallet?"),
                     _("Fee included:"), str_fee, NULL);
 }
 
+void layoutConfirmReplacement(const char *description, uint8_t txid[32]) {
+  const char **str = split_message_hex(txid, 32);
+  layoutDialogSwipe(&bmp_icon_question, _("Cancel"), _("Confirm"), NULL,
+                    description, str[0], str[1], str[2], str[3], NULL);
+}
+
+void layoutConfirmModifyFee(const CoinInfo *coin, uint64_t fee_old,
+                            uint64_t fee_new) {
+  char str_fee_change[32] = {0};
+  char str_fee_new[32] = {0};
+  char *question = NULL;
+
+  uint64_t fee_change = 0;
+  if (fee_old < fee_new) {
+    question = _("Increase your fee by:");
+    fee_change = fee_new - fee_old;
+  } else {
+    question = _("Decrease your fee by:");
```

Comment:
Originally I had a third branch "Your fee did not change" like we have on T2, but I removed it in the "final UI tweaks" f1ab2b0 shortly before you commented. Since the scenarios that can occur on T1 are simpler than those on T2, I decided not to show the `layoutConfirmModifyFee()` dialog at all if there is no fee change. So `Decrease your fee by 0` is unreachable. If the fee change is 0, then instead of showing "Modify fee for TXID: ..." on the first screen I decided to show "Update TXID: ...". The scenarios where someone would want to do that all seem a bit unlikely (see code [comment](https://github.com/trezor/trezor-firmware/commit/f1ab2b0582299475bf3ed23bf8f66657826a6c21#diff-6a32ff6ab464699431f3fd383c5dfc926fba103a967f3086852e84540b874ba7R1548-R1551)), but I don't see a reason to forbid it.

Reply:
We agreed at the standup that not showing `layoutConfirmModifyFee()` is alright, but I am leaving this comment unresolved for the moment just in case someone want's to discuss showing "Update TXID" in case of 0 fee.